File: src/dorkbox/network/dns/records/SingleNameBase.java
// Copyright (c) 2004 Brian Wellington (bwelling@xbill.org)

package dorkbox.network.dns.records;

import java.io.IOException;

import dorkbox.network.dns.Compression;
import dorkbox.network.dns.DnsInput;
import dorkbox.network.dns.DnsOutput;
import dorkbox.network.dns.Name;
import dorkbox.network.dns.utils.Tokenizer;

/**
 * Implements common functionality for the many record types whose format
 * is a single name.
 *
 * @author Brian Wellington
 */

abstract
class SingleNameBase extends DnsRecord {

    private static final long serialVersionUID = -18595042501413L;

    protected Name singleName;

    protected
    SingleNameBase() {}

    protected
    SingleNameBase(Name name, int type, int dclass, long ttl) {
        super(name, type, dclass, ttl);
    }

    protected
    SingleNameBase(Name name, int type, int dclass, long ttl, Name singleName, String description) {
        super(name, type, dclass, ttl);
        this.singleName = checkName(description, singleName);
    }

    @Override
    void rrFromWire(DnsInput in) throws IOException {
        singleName = new Name(in);
    }

    @Override
    void rrToWire(DnsOutput out, Compression c, boolean canonical) {
        singleName.toWire(out, null, canonical);
    }

    @Override
    void rrToString(StringBuilder sb) {
        sb.append(singleName.toString());
    }

    @Override
    void rdataFromString(Tokenizer st, Name origin) throws IOException {
        singleName = st.getName(origin);
    }

    protected
    Name getSingleName() {
        return singleName;
    }

}

File: src/dorkbox/dns/dns/records/SingleNameBase.kt
/*
 * Copyright 2021 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.dns.dns.records

import dorkbox.dns.dns.Compression
import dorkbox.dns.dns.DnsInput
import dorkbox.dns.dns.DnsOutput
import dorkbox.dns.dns.Name
import dorkbox.dns.dns.utils.Tokenizer
import java.io.IOException

/**
 * Implements common functionality for the many record types whose format
 * is a single name.
 *
 * @author Brian Wellington
 */
abstract class SingleNameBase : DnsRecord {
    var singleName: Name = Name.empty

    protected constructor()
    protected constructor(name: Name, type: Int, dclass: Int, ttl: Long) : super(name, type, dclass, ttl)
    protected constructor(name: Name, type: Int, dclass: Int, ttl: Long, singleName: Name, description: String) : super(name, type, dclass, ttl) {
        this.singleName = checkName(description, singleName)
    }

    @Throws(IOException::class)
    override fun rrFromWire(`in`: DnsInput) {
        singleName = Name(`in`)
    }

    override fun rrToWire(out: DnsOutput, c: Compression?, canonical: Boolean) {
        singleName.toWire(out, null, canonical)
    }

    override fun rrToString(sb: StringBuilder) {
        sb.append(singleName.toString())
    }

    @Throws(IOException::class)
    override fun rdataFromString(st: Tokenizer, origin: Name?) {
        singleName = st.getName(origin)
    }

    companion object {
        private const val serialVersionUID = -18595042501413L
    }
}
